core/src/services/hdfs_native/reader.rs

```diff
@@ -15,23 +15,58 @@
 // specific language governing permissions and limitations
 // under the License.
 
+use futures::StreamExt;
 use hdfs_native::file::FileReader;
 
 use crate::raw::*;
+use crate::services::hdfs_native::error::parse_hdfs_error;
 use crate::*;
 
 pub struct HdfsNativeReader {
-    _f: FileReader,
+    f: FileReader,
+    offset: usize,
+    read: usize,
+    size: usize,
+    buf_size: usize,
+    buf: Buffer,
 }
 
 impl HdfsNativeReader {
-    pub fn new(f: FileReader) -> Self {
-        HdfsNativeReader { _f: f }
+    pub fn new(f: FileReader, offset: usize, size: usize) -> Self {
+        HdfsNativeReader {
+            f,
+            offset,
+            read: 0,
+            size,
+            // Use 2 MiB as default value.
+            buf_size: 2 * 1024 * 1024,
+            buf: Buffer::new(),
+        }
     }
 }
 
 impl oio::Read for HdfsNativeReader {
     async fn read(&mut self) -> Result<Buffer> {
-        todo!()
+        if self.read >= self.size {
+            return Ok(Buffer::new());
+        }
+
+        if self.buf.is_empty() {
```

Comment:
I didn't get why we need this. Can we store the stream directly and avoid copying them?